numpy/core/src/multiarray/descriptor.c

```diff
@@ -1439,6 +1439,12 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
         PyObject *obj2;
         obj2 = PyUnicode_AsASCIIString(obj);
         if (obj2 == NULL) {
+            /* Convert the exception into a TypeError */
+            PyObject *err = PyErr_Occurred();
+            if (PyErr_GivenExceptionMatches(err, PyExc_UnicodeEncodeError)) {
+                PyErr_SetString(PyExc_TypeError,
```

Comment:
Doesn't the old error need to be cleared?

Reply:
I was following a pattern I saw in other places that do the same, like [in conversion_utils.c](https://github.com/numpy/numpy/blob/464f79eb1d05bf938d16b49da1c39a4e02506fa3/numpy/core/src/multiarray/conversion_utils.c#L920). 

Looking at the CPython sources, it seems this is by design. `PyErr_SetString(exc, value)` calls `PyErr_SetObject(exc, value)`, which calls `PyErr_Restore(exc, value, tb)`, which gets the current values into temporary variables, sets the new values, then `Py_XDECREF`s the old ones.

Reply:
OK, good to know, the official documentation doesn't cover that. Google does show this:

>The other thing to note is that if there are multiple calls to PyErr_SetString only the last one counts:

I don't know if that covers other ways of raising errors, but I assume your inspection of the CPython code covers that.